src/Avalonia.Visuals/CornerRadius.cs

```diff
@@ -26,49 +28,127 @@ public CornerRadius(double uniformRadius)
         {
             TopLeft = TopRight = BottomLeft = BottomRight = uniformRadius;
 
+            TopLeftRadiusX     = TopLeftRadiusY     = TopLeft;
+            TopRightRadiusX    = TopRightRadiusY    = TopRight;
+            BottomRightRadiusX = BottomRightRadiusY = BottomRight;
+            BottomLeftRadiusX  = BottomLeftRadiusY  = BottomLeft;
+
+            _isCircular = true;
         }
+
         public CornerRadius(double top, double bottom)
         {
-            TopLeft = TopRight = top;
+            TopLeft    = TopRight    = top;
             BottomLeft = BottomRight = bottom;
+
+            TopLeftRadiusX     = TopLeftRadiusY     = TopLeft;
+            TopRightRadiusX    = TopRightRadiusY    = TopRight;
+            BottomRightRadiusX = BottomRightRadiusY = BottomRight;
+            BottomLeftRadiusX  = BottomLeftRadiusY  = BottomLeft;
+
+            _isCircular = true;
         }
-        public CornerRadius(double topLeft, double topRight, double bottomRight, double bottomLeft)
+
+        public CornerRadius(
+            double topLeft,
+            double topRight,
+            double bottomRight,
+            double bottomLeft)
         {
-            TopLeft = topLeft;
-            TopRight = topRight;
+            TopLeft     = topLeft;
+            TopRight    = topRight;
             BottomRight = bottomRight;
-            BottomLeft = bottomLeft;
+            BottomLeft  = bottomLeft;
+
+            TopLeftRadiusX     = TopLeftRadiusY     = TopLeft;
+            TopRightRadiusX    = TopRightRadiusY    = TopRight;
+            BottomRightRadiusX = BottomRightRadiusY = BottomRight;
+            BottomLeftRadiusX  = BottomLeftRadiusY  = BottomLeft;
+
+            _isCircular = true;
+        }
+
+        public CornerRadius(
+            double topLeftRadiusX,
+            double topLeftRadiusY,
+            double topRightRadiusX,
+            double topRightRadiusY,
+            double bottomRightRadiusX,
+            double bottomRightRadiusY,
+            double bottomLeftRadiusX,
+            double bottomLeftRadiusY)
+        {
+            TopLeftRadiusX     = topLeftRadiusX;
+            TopLeftRadiusY     = topLeftRadiusY;
+            TopRightRadiusX    = topRightRadiusX;
+            TopRightRadiusY    = topRightRadiusY;
+            BottomRightRadiusX = bottomRightRadiusX;
+            BottomRightRadiusY = bottomRightRadiusY;
+            BottomLeftRadiusX  = bottomLeftRadiusX;
+            BottomLeftRadiusY  = bottomLeftRadiusY;
+
+            TopLeft     = (TopLeftRadiusX     + TopLeftRadiusY)     / 2.0;
+            TopRight    = (TopRightRadiusX    + TopRightRadiusY)    / 2.0;
+            BottomRight = (BottomRightRadiusX + BottomRightRadiusY) / 2.0;
+            BottomLeft  = (BottomLeftRadiusX  + BottomLeftRadiusY)  / 2.0;
```

Comment:
This is the only part that really can be disagreed with in my opinion. Some existing code expects circular corner radii and therefore something needs to be provided for backwards compatibility. In that case an average of X/Y components seems fair. It would be unintuitive for developers though so code paths need to be updated to handle elliptical cases correctly.